File: app/src/main/java/lazy/of/go/to/module/MainModule.java
package lazy.of.go.to.module;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;
import dagger.android.ContributesAndroidInjector;
import lazy.of.go.to.app.MainActivity;
import lazy.of.go.to.app.MainContract;
import lazy.of.go.to.app.MainFragment;
import lazy.of.go.to.app.MainPresenter;
import lazy.of.go.to.common.Log;
import lazy.of.go.to.di.ActivityScoped;
import lazy.of.go.to.di.FragmentScoped;

/**
 * @author piece.of.lazy
 */
@Module
public abstract class MainModule {

    @FragmentScoped
    @ContributesAndroidInjector
    abstract MainFragment injectFragment();

    @ActivityScoped
    @Binds
    abstract MainContract.Presenter injectPresenter(MainPresenter presenter);

    @ActivityScoped
    @Provides
    static Log provideLog(MainActivity activity) {
        return activity.getLog();
    }
}

File: app/src/main/kotlin/lazy/of/go/to/app/main/MainModule.kt
package lazy.of.go.to.app.main

import dagger.Binds
import dagger.Module
import dagger.Provides
import dagger.android.ContributesAndroidInjector
import lazy.of.go.to.base.feature.GetFeature
import lazy.of.go.to.common.Log
import lazy.of.go.to.db.DbInjection
import lazy.of.go.to.di.ActivityScoped
import lazy.of.go.to.di.FragmentScoped
import lazy.of.go.to.domain.data.SettingRefRep
import lazy.of.go.to.domain.data.UserRepository

/**
 * @author piece.of.lazy
 */
@Module
abstract class MainModule {
    @FragmentScoped
    @ContributesAndroidInjector
    internal abstract fun injectFragment(): MainFragment

    @ActivityScoped
    @Binds
    internal abstract fun injectPresenter(presenter: MainPresenter): MainContract.Presenter

    @Module
    companion object {
        @JvmStatic
        @ActivityScoped
        @Provides
        internal fun provideLog(activity: MainActivity): Log = activity.log

        @JvmStatic
        @ActivityScoped
        @Provides
        internal fun provideGetFeature(activity: MainActivity): GetFeature = activity

        @JvmStatic
        @ActivityScoped
        @Provides
        internal fun provideDbUser(dbInjection: DbInjection) = dbInjection.getDB(UserRepository::class)

        @JvmStatic
        @ActivityScoped
        @Provides
        internal fun provideSettingReference(dbInjection: DbInjection) = dbInjection.getDB(SettingRefRep::class)
    }
}